lib/spack/spack/build_systems/cached_cmake.py

```diff
@@ -0,0 +1,221 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+import os
+
+from spack.build_systems.cmake import CMakePackage
+from spack.package import run_after
+
+
+def cmake_cache_entry(name, value, comment=""):
+    """Generate a string for a cmake cache variable"""
+    return 'set({0} "{1}" CACHE PATH "{2}")\n'.format(name, value, comment)
+
+
+def cmake_cache_option(name, boolean_value, comment=""):
+    """Generate a string for a cmake configuration option"""
+
+    value = "ON" if boolean_value else "OFF"
+    return 'set({0} {1} CACHE BOOL "{2}")\n'.format(name, value, comment)
+
+
+class CachedCMakePackage(CMakePackage):
+    """Specialized class for packages build using CMake initial cache.
+
+    This feature of CMake allows packages to increase reproducibility,
+    especially between Spack- and manual builds. It also allows packages to
+    sidestep certain parsing bugs in extremely long ``cmake`` commands, and to
+    avoid system limits on the length of the command line."""
+
+    phases = ['initconfig', 'cmake', 'build', 'install']
+
+    @property
+    def cache_name(self):
+        return "{0}-{1}-{2}-{3}.cmake".format(
+            self.name,
+            self.spec.architecture,
+            self.spec.compiler.name,
+            self.spec.compiler.version,
+        )
+
+    @property
+    def cache_path(self):
+        return os.path.join(self.stage.source_path, self.cache_name)
+
+    def flag_handler(self, name, flags):
+        if name in ('cflags', 'cxxflags', 'cppflags', 'fflags'):
+            return (None, None, None)  # handled in the cmake cache
+        return (flags, None, None)
+
+    def initconfig_compiler_entries(self):
+        # This will tell cmake to use the Spack compiler wrappers when run
+        # through Spack, but use the underlying compiler when run outside of
+        # Spack
+        spec = self.spec
+        entries = [
+            "#------------------{0}".format("-" * 60),
+            "# Compilers",
+            "#------------------{0}\n".format("-" * 60),
+
+            "# Compiler Spec: {0}".format(spec.compiler),
+            "#------------------{0}".format("-" * 60),
+            'if(DEFINED ENV{SPACK_CC})',
+            '  ' + cmake_cache_entry("CMAKE_C_COMPILER", env['CC']),
+            '  ' + cmake_cache_entry("CMAKE_CXX_COMPILER", env['CXX']),
+            '  ' + cmake_cache_entry("CMAKE_Fortran_COMPILER", env['FC']),
+            'else()',
+            '  ' + cmake_cache_entry("CMAKE_C_COMPILER", spack_cc),
+            '  ' + cmake_cache_entry("CMAKE_CXX_COMPILER", spack_cxx),
+            '  ' + cmake_cache_entry("CMAKE_Fortran_COMPILER", spack_fc),
+            'endif()'
+        ]
+
+        # use global spack compiler flags
+        cppflags = ' '.join(spec.compiler_flags['cppflags'])
```

Comment:
Should these be moved into the `else` block right above?  They shouldn't be needed unless building outside of Spack, since they are handled by the Spack compiler wrappers...

Reply:
No, I added code in the `flag_handler` to ignore them in the Spack wrapper so they're only handled in the one place.

Reply:
:+1: